accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,632 @@
+
+Code on GitHub, GitLab etc can be browsed and audited, but packages in the registry are opaque and much harder to vet. If packages are built and published in the open these attacks become a whole lot harder. However, right now there’s no way of knowing where a package came from when you retrieve it from the registry.
+
+We want to solve the problem of npm packages being disconnected from their source by linking the published package to the source code repository and build that it originated from.
+
+Once a package includes provenance information (where and how it was built) we can start showing this when browsing a package on the npm registry. You would be able to click through to the particular commit and build that published a given version:
+
+<img width="376" alt="screenshot of npm package with provenance information" src="https://user-images.githubusercontent.com/20165/176637201-fdb02c11-c810-48e1-a203-ea4fb2008ad3.png">
+
+Developers consuming open source packages should get the benefit of this without any changes to their workflows. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.
+
+Open source maintainers must be able to link their packages to the source and build it originated from with near-zero initial and ongoing overhead.
+
+## Goals
+- Establishes a verifiable link between a public npm package and the source repository and build it originated from.
+- Does not expose any Personally identifiable information (PII) about maintainers, e.g. emails.
+- Avoids developer-managed keys (as there's no good way to offer trust to the community given the challenges of distributing public keys). This, along with opt-in signing, adds near-zero initial and ongoing overhead for open source maintainers.
+- Maintainers can opt-in to linking their packages to the source and build it originated from when publishing using `npm publish` for public packages.
+- Incentivizes maintainers to build in the open because of the strong guarantees that this offers.
+- Verification happens transparently on `npm install` without the need to obtain or manage additional tools or keys.
+- Verification should have a negligible performance impact on `npm install`.
+- Verification should be performed without depending on any third-party systems other than the registry.
+- Compatible with third-party npm clients, e.g. `yarn` and `pnpm`.
+- Should allow third-party npm registries, e.g. GitHub Packages, Artifactory and Verdaccio to follow suit and implement similar interfaces.
+- Should be maintained with >99.9% uptime so that developers are not blocked from publishing new packages.
+- Should allow future extensions to support centrally managed signing authorities such as certificates managed by an enterprise and inner source within air-gapped enterprise environments.
```

Comment:
It it possible to run a private rekor (transparency ledger that drives sigstore) instance and pair it up to SPIFFE/SPIRE to achieve this. Not sure how difficult this process is though. If this is an important use case, we can work on simplifying this path in the upstream projects.

Reply:
> It it possible to run a private rekor (transparency ledger that drives sigstore) instance and pair it up to SPIFFE/SPIRE to achieve this. Not sure how difficult this process is though. If this is an important use case, we can work on simplifying this path in the upstream projects.

Yes I have a vague understanding that something like this should be possible. It's not an immediate goal to support this use case, I'd like to make sure there is some path to get there in future that won't require major reworking.